### PR TITLE
ENH: adding documentation for refactored multi-threading infrastructure

### LINKED ISSUE #67: Update Section `3.2.7 Multi-Threading` to account for new ITK multi-threading
I think there have been some important changes to ITK's multi-threading classes lately, or that there is an important effort ongoing.

The section in `\SoftwareGuide\Latex\Architecture\SystemOverview.tex` is quite short, so may be some class name or some other minor addition may be required

@dzenanz ?

### PR DESCRIPTION
**Pull request by dzenanz**
Closes #67.

### CODE REVIEW

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 1]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
...
-In this example each thread invokes the same method. The multithreaded filter
-takes care to divide the image into different regions that do not overlap for
-write operations.
+In this example each thread invokes \code{DynamicThreadedGenerateData}
+method of the derived filter. The \code{ParallelizeImageRegion}
+method takes care to divide the image into different regions
+that do not overlap for write operations.
+\code{ImageSource}'s \code{GenerateData()} passes \code{this} pointer
+to \code{ParallelizeImageRegion}, which allows \code{ParallelizeImageRegion}
+to update filter's progress after each region is finished processing.
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `(...) to update the filter's progress (...)`?

File: SoftwareGuide/Latex/DevelopmentGuidelines/WriteAFilter.tex [hunk 2]
```diff
@@ -455,28 +455,33 @@ \section{Threaded Filter Execution}
...
 \code{cout} or \code{cerr}) or invoking events. A safe practice is to allow
 only thread id zero to perform I/O or generate events. (The thread id is
-passed as argument into \code{ThreadedGenerateData()}).  If more than one
-thread tries to write to the same place at the same time, the program can
-behave badly, and possibly even deadlock or crash.
+passed as argument into \code{ThreadedGenerateData(Region, ThreadId)}).
+If more than one thread tries to write to the same place at the same time,
+the program can behave badly, and possibly even deadlock or crash.
+
+Filters which need thread id should implement
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/DevelopmentGuidelines/WriteAFilter.tex]: `Filter which need thread identification should implement the (...)`

File: SoftwareGuide/Latex/DevelopmentGuidelines/WriteAFilter.tex [hunk 3]
```diff
@@ -444,7 +444,7 @@ \subsubsection{UpdateOutputData()}
 
 The developer will never override \code{UpdateOutputData()}. The developer need
 only write the \code{GenerateData()} method (non-threaded) or
-\code{ThreadedGenerateData()} method. A discussion of threading follows in the
+\code{DynamicThreadedGenerateData()} method. A discussion of threading follows in the
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/DevelopmentGuidelines/WriteAFilter.tex]: `A discussion on threading (...)`

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 4]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
...
+\doxygen{MultiThreaderBase}. And classes should use the multi-threaders via
+\code{MultiThreaderBase} interface, to allow the end user the flexibility to
+select the multi-threader at run time. This also allows the filter to
+benefit from future improvements in threading such as addition of a new
+multi-threader implementation.
+
+The backwards compatible \code{ThreadedGenerateData(Region, ThreadId)} method
+signature has been kept, for use in filters that must know their thread number.
+To use this signature, a filter must invoke
+\code{this->DynamicMultiThreadingOff();} prior to \code{Update();}. The best
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `(...) prior to \code{this->Update();}`

for consistency and conform to the style guidelines.

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 5]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
...
+place for that invocation is filter's constructor.
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `(...) is the filter's constructor.`

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 6]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
...
+in its own version of \code{GenerateData()}, such as done by
+\doxygen{CannyEdgeDetectionImageFilter}:
+
+\small
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{cpp}
+::GenerateData()
+{
+  this->UpdateProgress(0.0f);
+  Superclass::AllocateOutputs();
+  //small serial section
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `// Small serial section`

for consistency with the Style guidelines.

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 7]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
...
+  this->UpdateProgress(0.45f);
+
+  // Calculate the gradient of the second derivative
+  this->GetMultiThreader()->template ParallelizeImageRegion<TOutputImage::ImageDimension>(
+      this->GetOutput()->GetRequestedRegion(),
+      [this](const OutputImageRegionType & outputRegionForThread)
+      { this->ThreadedCompute2ndDerivativePos(outputRegionForThread); }, nullptr);
+  this->UpdateProgress(0.9f);
+
+  //more processing
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `// More processing`

for consistency with the Style guidelines.

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 8]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
 Multithreading is handled in ITK through a high-level design
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `Multi-threading`

for consistency. But I'm fine if we choose to remove the dash in **all** appearances.

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 9]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
...
 abstraction. This approach provides portable multithreading and hides the
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `multi-threading`

for consistency.

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex [hunk 10]
```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
...
 complexity of differing thread implementations on the many systems supported
-by ITK. For example, the class \doxygen{MultiThreader} provides support for
-multithreaded execution using \code{sproc()} on an SGI, or
-\code{pthread\_create} on any platform supporting POSIX threads.
+by ITK. For example, the class \doxygen{PlatformMultiThreader} provides support for
+multithreaded execution by directly using platform-specific primitives such as
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Architecture/SystemOverview.tex]: `multi-threaded`

for consistency.

File: SoftwareGuide/Latex/DevelopmentGuidelines/WriteAFilter.tex [hunk 11]
```diff
@@ -455,28 +455,33 @@ \section{Threaded Filter Execution}
 Filters that can process data in pieces can typically multi-process
 using the data parallel, shared memory implementation built into the
 pipeline execution process. To create a multithreaded filter, simply
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/DevelopmentGuidelines/WriteAFilter.tex]: `multi-threaded`

for consistency.

File: SoftwareGuide/Latex/Architecture/SystemOverview.tex (hunk 4 above)
Comment (dzenanz) [reply SoftwareGuide/Latex/Architecture/SystemOverview.tex]: Filters don't usually invoke Update() on themselves, rather the user of the filter does that. Hence no `this->` prefix. I will reword, though.

**Comment by dzenanz** [comment]
Thanks for review Jon! I updated the commit.

**Pull request merged**